internal/fields/mappings.go

```diff
@@ -0,0 +1,577 @@
+			previewValue := mappingParameter("value", preview)
+			if previewValue != actualValue {
+				// This should also be detected by the failure storage (if available)
+				// or no documents being ingested
+				return multierror.Error{
+					fmt.Errorf("constant_keyword value in preview %q does not match the actual mapping value %q for path: %q", previewValue, actualValue, path),
+				}
+			}
+			return nil
+		}
+		errs = append(errs, fmt.Errorf("invalid type for %q: no constant_keyword type set in preview mapping", path))
+		if len(errs) == 0 {
+			return nil
+		}
+		return errs.Unique()
+	}
+
+	if isObjectDynamic(actual) {
+		logger.Debugf("Dynamic object found but no fields ingested under path: %s.*", path)
+		return errs.Unique()
+	}
+
+	if isObject(actual) {
+		if isObjectDynamic(preview) {
+			// TODO: Skip for now, it should be required to compare with dynamic templates
+			logger.Debugf("Pending to validate with the dynamic templates defined the path: %s", path)
+			return errs.Unique()
+		} else if !isObject(preview) {
+			errs = append(errs, fmt.Errorf("not found properties in preview mappings for path: %s", path))
+			return errs.Unique()
+		}
+		previewProperties, err := getMappingDefinitionsField("properties", preview)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("found invalid properties type in preview mappings for path %q: %w", path, err))
+		}
+		actualProperties, err := getMappingDefinitionsField("properties", actual)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("found invalid properties type in actual mappings for path %q: %w", path, err))
+		}
+		// logger.Debugf(">>> Comparing field with properties (object): %q", path)
+		compareErrors := compareMappings(path, mappingDefinitions(previewProperties), mappingDefinitions(actualProperties), ecsSchema, localSchema)
+		errs = append(errs, compareErrors...)
+
+		if len(errs) == 0 {
+			return nil
+		}
+		return errs.Unique()
+	}
+
+	if isMultiFields(actual) {
+		if !isMultiFields(preview) {
+			errs = append(errs, fmt.Errorf("not found multi_fields in preview mappings for path: %s", path))
+			return errs.Unique()
+		}
+		previewFields, err := getMappingDefinitionsField("fields", preview)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("found invalid multi_fields type in preview mappings for path %q: %w", path, err))
+		}
+		actualFields, err := getMappingDefinitionsField("fields", actual)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("found invalid multi_fields type in actual mappings for path %q: %w", path, err))
+		}
+		// logger.Debugf(">>> Comparing multi_fields: %q", path)
+		compareErrors := compareMappings(path, mappingDefinitions(previewFields), mappingDefinitions(actualFields), ecsSchema, localSchema)
+		errs = append(errs, compareErrors...)
+		// not returning here to keep validating the other fields of this object if any
+	}
+
+	// Compare all the other fields
+	containsMultifield := isMultiFields(actual)
+	for key, value := range actual {
```

Comment:
There could be more objects at this point. Here it is evaluating/validating all the elements under the `properties` field (or the initial elements in the mappings).

So, for instance in these mappings:
```json
  "apache": {
    "properties": {
      "access": {
        "properties": {
          "http": {
            "properties": {
              "request_headers": {
                "type": "keyword",
                "ignore_above": 1024
              }
            }
          },
          "identity": {
            "type": "keyword",
            "ignore_above": 1024
          },
          "remote_addresses": {
            "type": "keyword",
            "ignore_above": 1024
          },
        }
      }
    }
  }
```

It would be checking:
- `apache` (as the first level element)
- `apache.access` (element under `apache` properties)
- `apache.access.http` (element under `apache.access` properties)
- `apache.access.http.request_headers` ...
- `apache.access.identity` ...
- `apache.access.remote_addresses` ...

So, in general, here it is validating the elements found within the `properties` field (fields or nested objects).

And at the same time, there could be elements here that do not have the corresponding definition in the preview (type array, defined in ECS or part of dynamic templates).

For those cases that could came from ECS or from dynamic templates, there could also be more objects under this path.

Reply:
> Seems complex to have to flatten all their properties and compare them one by one.

This flatten process takes place for those elements that does not exist in the preview mapping. Those elements/fields could be defined in the ECS, matched with some dynamic template or being of type `array`.